File: blaze_flutter_sdk/android/src/main/kotlin/com/blaze/flutterblazesdk/utils/BlazePlayerEventData.kt
package com.blaze.flutterblazesdk.utils

import android.util.Log
import androidx.annotation.Keep
import com.blaze.blazesdk.delegates.models.BlazePlayerEvent
import com.blaze.flutterblazesdk.utils.parsing.gson.toJsonString

/**
 * Data class that wraps BlazePlayerEvent data for native-to-Flutter communication This centralizes
 * the player event data structure and can be reused across the SDK
 *
 * Usage example:
 * ```kotlin
 * val eventData = BlazePlayerEventData.fromBlazePlayerEvent(playerEvent)
 * // Use eventData.toJsonString() to send to Flutter
 * ```
 */
@Keep
data class BlazePlayerEventData(val playerEventType: String, val playerEventParams: Any) {

    companion object {
        /**
         * Create BlazePlayerEventData from a native BlazePlayerEvent This is the main factory
         * method for converting native events
         */
        fun fromBlazePlayerEvent(event: BlazePlayerEvent): BlazePlayerEventData {
            val eventType = event.toFlutterEventType()
            val eventParams = event.toFlutterEventParams()

            return BlazePlayerEventData(
                    playerEventType = eventType,
                    playerEventParams = eventParams
            )
        }

    }

}

/// Extension function to convert BlazePlayerEvent to Flutter event type string
fun BlazePlayerEvent.toFlutterEventType(): String {
    return when (this) {
        is BlazePlayerEvent.OnMomentStart -> "OnMomentStart"
        is BlazePlayerEvent.OnStoryStart -> "OnStoryStart"
        is BlazePlayerEvent.OnVideoStart -> "OnVideoStart"
    }
}

/// Extension function to convert BlazePlayerEvent parameters to JSON string
fun BlazePlayerEvent.toFlutterEventParams(): Any {
    val params: Any =
        when (this) {
            is BlazePlayerEvent.OnMomentStart -> {
                @Keep
                data class OnMomentStartParams(val momentId: String)
                OnMomentStartParams(momentId = this.params.momentId)
            }
            is BlazePlayerEvent.OnStoryStart -> {
                @Keep
                data class OnStoryStartParams(val storyId: String)
                OnStoryStartParams(storyId = this.params.storyId)
            }
            is BlazePlayerEvent.OnVideoStart -> {
                @Keep
                data class OnVideoStartParams(val videoId: String)
                OnVideoStartParams(videoId = this.params.videoId)
            }
        }

    return params
}

File: blaze_flutter_sdk/ios/blaze_flutter_sdk/Sources/Utils/BlazePlayerEventData.swift
//
//  BlazePlayerEventData.swift
//  blaze_flutter_sdk
//
//  Created by Created by Reuven Levitsky.
//

import Foundation
import BlazeSDK

/**
 * Codable struct that wraps BlazePlayerEvent data for native-to-Flutter communication
 * This centralizes the player event data structure and can be reused across the SDK
 *
 * Usage example:
 * ```swift
 * let eventData = BlazePlayerEventData.from(playerEvent)
 * // Use eventData for Flutter communication
 * ```
 */
struct BlazePlayerEventData: Encodable {
    let playerEventType: String
    let playerEventParams: AnyEncodable

    static func from(_ event: BlazePlayerEvent) -> BlazePlayerEventData {
        let eventType = event.toFlutterEventType
        let eventParams = event.toFlutterEventParams

        return BlazePlayerEventData(
            playerEventType: eventType,
            playerEventParams: AnyEncodable(eventParams)
        )
    }
}

extension BlazePlayerEvent {
    
    /// Convert BlazePlayerEvent to Flutter event type string
    var toFlutterEventType: String {
        switch self {
        case .onMomentStart:
            return "OnMomentStart"
        case .onStoryStart:
            return "OnStoryStart"
        case .onVideoStart:
            return "OnVideoStart"
        }
    }
    
    /// Convert BlazePlayerEvent parameters to JSON string
    var toFlutterEventParams: Encodable {
        switch self {
        case .onMomentStart(let params):
            struct Params: Codable {
                let momentId: String
            }
            return Params(momentId: params.momentId)
        case .onStoryStart(let params):
            struct Params: Codable {
                let storyId: String
            }
            return Params(storyId: params.storyId)
        case .onVideoStart(let params):
            struct Params: Codable {
                let videoId: String
            }
            return Params(videoId: params.videoId)
        }
    }
}
